opm/simulators/linalg/bda/rocsparseWellContributions.cpp

```diff
@@ -0,0 +1,239 @@
+                localSum[wiId] += localSum[wiId + i*valsPerBlock];
+	    }
+        }
+
+        if(c == 0 && wiId < valsPerBlock){
+            for(unsigned i = dim - 1; i > 0; --i){
+                localSum[wiId] += localSum[wiId + i];
+            }
+            z1[r] = localSum[wiId];
+        }
+    }
+
+    __syncthreads();
+
+    if(wiId < dim_wells){
+        temp = 0.0;
+        for(unsigned i = 0; i < dim_wells; ++i){
+            temp += Dnnzs[wgId*dim_wells*dim_wells + wiId*dim_wells + i]*z1[i];
+        }
+        z2[wiId] = temp;
+    }
+
+    __syncthreads();
+
+    if(wiId < dim*valSize){
+        temp = 0.0;
+        unsigned bb = wiId/dim + val_pointers[wgId];
+        for (unsigned j = 0; j < dim_wells; ++j){
+            temp += Cnnzs[bb*dim*dim_wells + j*dim + c]*z2[j];
+        }
+
+        int colIdx = Ccols[bb];
+        y[colIdx*dim + c] -= temp;
+    }
+}
+#endif
+
+
+void WellContributionsRocsparse::apply_stdwells([[maybe_unused]] double *d_x,
+                                                [[maybe_unused]] double *d_y){
+#ifdef __HIP__
+    unsigned gridDim = num_std_wells;
+    unsigned blockDim = 32;
+    unsigned shared_mem_size = (blockDim + 2 * dim_wells) * sizeof(double); // shared memory for localSum, z1 and z2
+    // dim3(N) will create a vector {N, 1, 1}
+    stdwell_apply<<<dim3(gridDim), dim3(blockDim), shared_mem_size, stream>>>(
+        d_Cnnzs_hip, d_Dnnzs_hip, d_Bnnzs_hip, d_Ccols_hip, d_Bcols_hip,
+        d_x, d_y, dim, dim_wells, d_val_pointers_hip
+    );
+    HIP_CHECK(hipStreamSynchronize(stream));
+#else
+    OPM_THROW(std::logic_error, "Error separate wellcontributions for rocsparse only supported when compiling with hipcc");
+#endif
+}
+
+void WellContributionsRocsparse::apply_mswells(double *d_x, double *d_y){
+    if (h_x.empty()) {
+        h_x.resize(N);
+        h_y.resize(N);
+    }
+
+    HIP_CHECK(hipMemcpyAsync(h_x.data(), d_x, sizeof(double) * N, hipMemcpyDeviceToHost, stream));
+    HIP_CHECK(hipMemcpyAsync(h_y.data(), d_y, sizeof(double) * N, hipMemcpyDeviceToHost, stream));
+    HIP_CHECK(hipStreamSynchronize(stream));
+
+    // actually apply MultisegmentWells
+    for (auto& well : multisegments) {
+        well->apply(h_x.data(), h_y.data());
+    }
+
+    // copy vector y from CPU to GPU
+    HIP_CHECK(hipMemcpyAsync(d_y, h_y.data(), sizeof(double) * N, hipMemcpyHostToDevice, stream));
+    HIP_CHECK(hipStreamSynchronize(stream));
+}
+
+void WellContributionsRocsparse::apply(double *d_x, double *d_y){
+    if(num_std_wells > 0){
+        apply_stdwells(d_x, d_y);
+    }
+
+    if(num_ms_wells > 0){
+        apply_mswells(d_x, d_y);
+    }
+}
+
+void WellContributionsRocsparse::setStream(hipStream_t stream_){
+    stream = stream_;
+}
+
+void WellContributionsRocsparse::APIaddMatrix(MatrixType type,
+                                        int* colIndices,
+                                        double* values,
+                                        unsigned int val_size)
```

Comment:
Would be cool if we could fix indentation here,

Reply:
Done